wasmplugin/body.go

```diff
@@ -0,0 +1,50 @@
+// Copyright The OWASP Coraza contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package wasmplugin
+
+import (
+	"io"
+
+	"github.com/tetratelabs/proxy-wasm-go-sdk/proxywasm"
+)
+
+type ByteLenger interface {
+	Len() int
+}
+
+// bodyWrapper is a wrapper around the GetHttpRequestBody which exposes
+// the length of the body on beforehand so the tx.ReadRequestBodyFrom can
+// look at the body length first before attempting to copy the bytes.
+type bodyWrapper struct{ totalSize int }
+
+var (
+	_ io.Reader  = bodyWrapper{}
+	_ ByteLenger = bodyWrapper{}
+)
+
+func (w bodyWrapper) Read(p []byte) (int, error) {
```

Comment:
This seems OK as long as `Read` is only called once to reduce wasm host function calls - with all the `io.*` optional interfaces it's not so clear if that's the case but if we can confirm it the complexity could be worth it. It is fragile though, see below